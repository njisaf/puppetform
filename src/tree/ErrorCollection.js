import { types } from 'mobx-state-tree';

const ErrorCollection = types.model('ErrorCollection', {
    failures: types.map(types.frozen())
})
.actions(self => {
    return {
        failure(error, value) {
            console.log('error! ', error, value);
            self.failures.set(error, value);
        }
    }
})

export default ErrorCollection;