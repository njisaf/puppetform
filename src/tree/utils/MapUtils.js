import { types } from 'mobx-state-tree';

const MapUtils = types.model()
.named('MapUtils')
.actions(self => {
    return {
        set(key, value) {
            try {
                self.map.set(key, value);
            } catch (e) {
                self.failure(key, value);
            }
        },
        delete(key) {
            self.map.delete(key);
        },
        clear() {
            self.map.clear()
        },
    }
})
.views(self => {
    return {
        get size() {
            return self.map.size;
        },
        get keys() {
            return self.map.keys();
        },
        get values() {
            return self.map.values();
        },
        get entries() {
            return self.map.entries();
        },
        has(key) {
            return self.map.has(key);
        },
        get(key) {
            return self.map.get(key);
        }
    }
})

export default MapUtils;