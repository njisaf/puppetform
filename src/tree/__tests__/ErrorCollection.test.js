import {
    getSnapshot,
    onSnapshot,
    types
} from 'mobx-state-tree';
import ErrorCollection from '../ErrorCollection';
import MapUtils from '../utils/MapUtils';

describe('ErrorCollection', () => {
    let states = [];

    const TestModel = types.compose(ErrorCollection, MapUtils, types.model({
        identity: types.string,
        alive: true,
    }).actions(self => ({
        kill() {
            self.alive = false;
            self.failure(self.identity, self.alive);
        }
    })))

    const testModel = TestModel.create({
        identity: 'testOne'
    });
  
    onSnapshot(testModel, snapshot => {
        states.push(snapshot)
    })


    it('testModel.alive can toggle to false', () => {
        console.log('testModel', getSnapshot(testModel))
        expect(testModel.alive).toBe(true);
        testModel.kill();
        expect(testModel.alive).toBe(false);
    })

    it('has the dead model added to the failures map on ErrorCollection', () => {
        expect(testModel.failures.size).toBe(1);
        expect(testModel.failures.get('testOne')).toBe(false);
    })

    // const testModel2 = TestModel.create({
    //     identity: 'testTwo'
    // })

    // it('adds a second dead model to the same ErrorCollection instane', () => {
    //     testModel2.kill();
    //     console.log('testModel2', getSnapshot(testModel2))
    //     expect(testModel2.failures.size).toBe(2);
    // })

    it('matches snapshot', () => {
        expect(states).toMatchSnapshot();
    })
})