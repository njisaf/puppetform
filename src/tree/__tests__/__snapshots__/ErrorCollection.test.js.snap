// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorCollection matches snapshot 1`] = `
Array [
  Object {
    "alive": false,
    "failures": Object {
      "testOne": false,
    },
    "identity": "testOne",
  },
]
`;
